extra/dns/hostctl.nix

```diff
@@ -0,0 +1,62 @@
+{ lib, pkgs, config, ... }:
+with lib;
+let
+  cfg = config.hostctl;
+  profile = toLower config.devshell.name;
+
+  etcHosts = pkgs.writeText "${profile}-etchosts" (
+    concatStringsSep "\n"
+      (mapAttrsToList (host: ip: ip + " " + host) cfg.dns)
+  );
+
+  # Execute this script to install the project's static dns entries
+  install-hostctl-dns = pkgs.writeShellScriptBin "install-hostctl-dns" ''
+    set -euo pipefail
+    shopt -s nullglob
+
+    log() {
+      IFS=$'\n' loglines=($*)
+      for line in ${"$"}{loglines[@]}; do echo -e "[hostctl] $line" >&2; done
+    }
+
+    # Install local CA into system, java and nss (includes Firefox) trust stores
+    log "Update static dns entries..."
+    sudo -K
+    log $(sudo ${pkgs.hostctl}/bin/hostctl add ${profile} --from ${etcHosts} 2>&1)
+
+    uninstall() {
+      log $(sudo ${pkgs.hostctl}/bin/hostctl remove ${profile} 2>&1)
+    }
+
+    # TODO: Uninstall when leaving the devshell
+    # trap uninstall EXIT
```

Comment:
This seems hard to get right, particularly because when the user runs two dev-shells and exits one.

Modifying the system is a bit heavy-handed for a development shell. I was secretly hoping for an unprivileged solution when I saw this PR. Perhaps (userns) containers are a more appropriate solution for most use cases.

Reply:
There is indeed a long term heavily politically involved solution to this: https://github.com/apetresc/nss-hostslocal. That would require lobbying for changing the way how linux system generally resolve hostfiles these days.

This particular PR is not going to get merged for this reason.

However, in a dev team setting, where you can exert certain powers over an individual's machine, this is a solution that (somewhat) works.